src/v/cluster/topic_recovery_service.cc

```diff
@@ -222,35 +222,48 @@ topic_recovery_service::recovery_status_log() const {
     return {_status_log.begin(), _status_log.end()};
 }
 
-static ss::future<result<std::vector<remote_segment_path>, recovery_error_ctx>>
-collect_manifest_paths(
+// NOTE rewritten as continuations to address arm64 miscompilation of coroutines
+// under clang-14
+static ss::future<std::vector<remote_segment_path>> collect_manifest_paths(
   remote& remote, ss::abort_source& as, const recovery_task_config& cfg) {
-    const auto& bucket = cfg.bucket;
-    auto rtc = make_rtc(as, cfg);
-
     // Look under each manifest prefix for topic manifests.
-    const char hex_chars[] = "0123456789abcdef";
-    std::vector<remote_segment_path> paths;
-    for (int i = 0; i < 16; ++i) {
-        const auto prefix = fmt::format("{}0000000/", hex_chars[i]);
-        auto rtc = make_rtc(as, cfg);
-
-        // This request is restricted to prefix, it should only return the
-        // metadata files for a topic.
-        auto meta = co_await remote.list_objects(
-          bucket, rtc, cloud_storage_clients::object_key{prefix});
-        if (meta.has_error()) {
-            vlog(cst_log.error, "Failed to list meta items: {}", meta.error());
-            continue;
-        }
-
-        for (auto&& item : meta.value().contents) {
-            vlog(cst_log.trace, "adding path {} for {}", item.key, prefix);
-            paths.emplace_back(item.key);
-        }
-    }
-
-    co_return paths;
+    constexpr static auto hex_chars = std::string_view{"0123456789abcdef"};
```

Comment:
Thanks for picking this up!

I wonder if making this constexpr static is enough to avoid the miscompilation, though happy to keep this as is for now

Reply:
I thought so, but i don't know a good way to quickly check so i played safe. Do people fire a arm ci run and decompile it or do they use compiler explored and a good dose of shims?